openedx/core/djangoapps/course_date_signals/tests.py

```diff
@@ -40,3 +42,42 @@ def test_hidden_sections(self):
             actual = [(idx, section.display_name, offset) for (idx, section, offset) in utils.spaced_out_sections(self.course)]
 
         self.assertEqual(actual, expected_sections)
+
+    def test_dates_for_ungraded_assignments(self):
+        """
+        _has_assignment_blocks should return true if the argument block
+        children leaf nodes include an assignment that is graded and scored
+        """
+        with modulestore().bulk_operations(self.course.id):
+            sequence = ItemFactory(parent=self.course, category="sequential")
+            vertical = ItemFactory(parent=sequence, category="vertical")
+            sequence = modulestore().get_item(sequence.location)
+            self.assertEqual(_has_assignment_blocks(sequence), False)
+
+            # Ungraded problems do not count as assignment blocks
+            ItemFactory.create(
+                parent=vertical,
+                category='problem',
+                graded=True,
+                weight=0,
+            )
+            sequence = modulestore().get_item(sequence.location)
```

Comment:
this only seems to work if I reassign sequence using modulestore().get_item after every modification.
otherwise, when the called method gets the children of the sequence, the returned vertical doesn't have any children. 
not too familiar with how modulestore works so not sure how to fix this (I imagine it is possible to do this without these calls)

Reply:
Yeah, I'd expect you to need to reload after making a change (xblock caches things internally).

I'm a bit worried about the current logic, though. In particular, I think there's an issue where if you have both an ungraded problem and a graded problem in the same sequence, the sequence will be marked as having a date, but the ungraded problem won't be excluded from that date (by setting its `due` attribute explicitly to `None`).

I don't recall if we have any tests of `_gather_graded_items`, but it seems like it would be worth testing that method (or a caller of that method) and verifying that it sets ungraded problems in graded sequences to have a `None` `due` attribute.

Reply:
as far as I can tell locally it does set the due date to None (because _gather_graded_items goes through the sequence, the vertical and then the problem at which point `due if _has_assignment_blocks(next_item) else None`
returns None
(I think you can actually see this behavior being fixed in the screenshots I had posted in the description)

I don't think we have tests of that method but I can still write a test like you're suggesting